touchforms/backend/test_form_player_api.py

```diff
@@ -0,0 +1,233 @@
+        )
+
+        model = new_xform_session.fem
+        controller = new_xform_session.fec
+
+        def answer_assert(answer, index):
+            new_xform_session.answer_question(answer, index)
+            old_xform_session.answer_question(answer, index)
+            assert(json_equals(old_xform_session.walk(), new_xform_session.walk()))
+
+        assert(json_equals(old_xform_session.walk(), json.loads(WalkJson.walkToJson(model, controller))))
+
+        answer_assert("1", "1,1")
+        answer_assert("William", "1,4")
+        answer_assert("1", "3")
+        answer_assert("2", "5")
+        answer_assert("2", "5")
+        answer_assert("2", "3")
+        answer_assert("1", "5")
+
+        old_submit = xformplayer.submit_form(old_xform_session,)
+
+    def test_nested_walking(self):
+
+        xform_session = xformplayer.XFormSession(
+            self.nested_xform,
+            instance=None,
+            use_api=True,
+            **self.session_metadata
+        )
+
+        model = xform_session.fem
+        controller = xform_session.fec
+
+        assert(json_equals(xform_session.walk(), json.loads(WalkJson.walkToJson(model, controller))))
+
+    def test_form_answering(self):
+        old_xform_session = xformplayer.XFormSession(
+            self.xform,
+            instance=None,
+            use_api=False,
+            **self.session_metadata
+        )
+
+        new_xform_session = xformplayer.XFormSession(
+            self.xform,
+            instance=None,
+            use_api=True,
+            **self.session_metadata
+        )
+
+        def answer_assert(answer, index):
+            new_answer = new_xform_session.answer_question(answer, index)
+            old_answer = old_xform_session.answer_question(answer, index)
+            assert(json_equals(old_answer, new_answer))
+            assert(json_equals(old_xform_session.walk(), new_xform_session.walk()))
+
+
+        answer_assert("123 ABC", "1")
+        answer_assert("123", "2")
+        answer_assert("2.5", "3")
+        answer_assert("1970-10-23", "4")
+        #answer_assert("1970-10-23T12:30:30", "5") datetime not supported
+        answer_assert("12:30:30", "6")
+        answer_assert("william s pride", "7")
+        answer_assert("12345", "8")
+        answer_assert("00012345", "9")
+        answer_assert("2", "10")
+        answer_assert("1 2 3", "11")
+
+
+def compare_dict(first, second):
+    for key in first:
+        if not key in second:
+            print "Comparing KEY: ", key
+            print "First: ", first
+            print "===="
+            print "Second: ", second
+            print ("Second not has: ", key)
+            return False
+        if not json_equals(first.get(key), second.get(key)):
+            print "Comparing KEY: ", key
+            print "First: ", first.get(key)
+            print "===="
+            print "Second: ", second.get(key)
+            return False
+    return True
+
+def compare_list(first, second):
```

Comment:
in general python has functions for this: https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertEqual